changefeedccl: fix default retry config parsing for webhook sink

spiffyy99 opened the pull request:
Fixes bug where HTTP requests would be retried indefinitely if
no webhook sink config was provided. Now, max retries cannot be
set as zero, "inf" should be used instead.

Release justification: Essential fix to webhook sink which is a
major component of changefeeds in 21.2.

Resolves #69325

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69330)
<!-- Reviewable:end -->

miretskiy reviewed (approved):
> Reviewed 1 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69330)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng)

___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 2873 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69330#-Mhyimz6146eJHKpfUxY:-Mhyimz6146eJHKpfUxZ:b-8rt74f) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f536dbba5e0993458c12f21d64494ab3a227870/pkg/ccl/changefeedccl/changefeed_test.go#L2873)):*
> ```Go
> 		`CREATE CHANGEFEED FOR foo INTO $1 WITH webhook_sink_config='{"Retry": {"Max": 0}}'`,
> 		`webhook-https://fake-host`,
> 	)
> ```

add `Max: -1`?
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 4294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69330#-Mhyib1EBCvS2G9JB-Xy:-Mhyib1EBCvS2G9JB-Xz:b-fazp8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f536dbba5e0993458c12f21d64494ab3a227870/pkg/ccl/changefeedccl/changefeed_test.go#L4294)):*
> ```Go
> 		// check that running status correctly updates with retryable error
> 		testutils.SucceedsSoon(t, func() error {
> 			status, err := feedJob.FetchRunningStatus()
> ```

This is fine, but just want to let you know we have a way of intercepting those retryable errors via changefeed testing knobs (HandleRetryableError).
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 147 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69330#-MhygVQxAv8KRmH5VMne:-MhygVQxAv8KRmH5VMnf:bxyxj5x) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f536dbba5e0993458c12f21d64494ab3a227870/pkg/ccl/changefeedccl/sink_webhook.go#L147)):*
> ```Go
> 	var i int
> 	// try to parse as int
> 	if err := json.Unmarshal(b, &i); err != nil {
> ```

perhaps `strconv.ParseInt("str", 10, 64)` instead of unmarshal?
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 147 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69330#-Mhyhf2E1k0P77-SUp0T:-Mhyhf2E1k0P77-SUp0U:bt9sc2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f536dbba5e0993458c12f21d64494ab3a227870/pkg/ccl/changefeedccl/sink_webhook.go#L147)):*
> ```Go
> 	var i int
> 	// try to parse as int
> 	if err := json.Unmarshal(b, &i); err != nil {
> ```

Would it make sense to invert this if statement `err == nil` so that we don't have deep nested if/else structures?
```
if err := json.Unmarsha(); err == nil {
    // Got a number... Check it's > 0
     return
}

// Try to see if it's == "inf"
```

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69330#-Mhyg_LS2aMXucSONO0v:-Mhyg_LS2aMXucSONO0w:bboildy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f536dbba5e0993458c12f21d64494ab3a227870/pkg/ccl/changefeedccl/sink_webhook.go#L150)):*
> ```Go
> 		var s string
> 		// if that fails, try to parse as string (only accept 'inf')
> 		if err := json.Unmarshal(b, &s); err != nil {
> ```

again, we can just use equality...I think you should just be able to do strings.ToLower(string(b))?
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 153 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69330#-MhyhPXrEz1zqldCRP8T:-MhyhPXrEz1zqldCRP8U:bpj4puh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f536dbba5e0993458c12f21d64494ab3a227870/pkg/ccl/changefeedccl/sink_webhook.go#L153)):*
> ```Go
> 			return err
> 		}
> 		if strings.ToLower(s) == "inf" {
> ```

please add comment why 0... 


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69330)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @spiffyyeng)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69330#-Mhyg_LS2aMXucSONO0v:-Mi1pUGC9TrSPYoBf3l9:bxl0ao5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f536dbba5e0993458c12f21d64494ab3a227870/pkg/ccl/changefeedccl/sink_webhook.go#L150)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

again, we can just use equality...I think you should just be able to do strings.ToLower(string(b))?
</blockquote></details>

unmarshal removes the quotes around string though, it'll be parsed as "inf" with quotes if we do this.

it might be better to just unmarshal for clarity here as opposed to removing quotes ourselves, i'll leave a comment explaining this. 


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69330)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69330#-Mhyg_LS2aMXucSONO0v:-Mi1rT76ELR6u6ZxtYPx:bzganrm) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f536dbba5e0993458c12f21d64494ab3a227870/pkg/ccl/changefeedccl/sink_webhook.go#L150)):*
<details><summary><i>Previously, spiffyyeng (Ryan Min) wrote…</i></summary><blockquote>

unmarshal removes the quotes around string though, it'll be parsed as "inf" with quotes if we do this.

it might be better to just unmarshal for clarity here as opposed to removing quotes ourselves, i'll leave a comment explaining this. 
</blockquote></details>

I see...  Makes sense.  


<!-- Sent from Reviewable.io -->

spiffyy99 committed:
changefeedccl: fix default retry config parsing for webhook sink

Fixes bug where HTTP requests would be retried indefinitely if
no webhook sink config was provided. Now, max retries cannot be
set as zero, "inf" should be used instead.

Release justification: Essential fix to webhook sink which is a
major component of changefeeds in 21.2.

Resolves #69325

Release note: None
